Zotero/Controllers/IdentifierLookupController.swift

```diff
@@ -0,0 +1,445 @@
+                    let lookupWebViewHandler = LookupWebViewHandler(lookupSettings: lookupSettings, webView: webView, translatorsController: self.translatorsController)
+                    self.lookupWebViewHandlersByLookupSettings[lookupSettings] = lookupWebViewHandler
+                }
+            }
+            guard let lookupWebViewHandler = self.lookupWebViewHandlersByLookupSettings[lookupSettings] else {
+                DDLogError("IdentifierLookupController: can't create LookupWebViewHandler instance")
+                return
+            }
+            self.setupObserver(for: lookupWebViewHandler)
+            lookupData = self.lookupData
+        }
+    }
+    
+    func lookUp(libraryId: LibraryIdentifier, collectionKeys: Set<String>, identifier: String) {
+        let lookupSettings = LookupWebViewHandler.LookupSettings(libraryIdentifier: libraryId, collectionKeys: collectionKeys)
+        lookupWebViewHandlersByLookupSettings[lookupSettings]?.lookUp(identifier: identifier)
+    }
+    
+    func cancelAllLookups() {
+        accessQueue.async(flags: .barrier) { [weak self] in
+            guard let self else { return }
+            DDLogInfo("IdentifierLookupController: cancel all lookups")
+            let keys = self.lookupWebViewHandlersByLookupSettings.keys
+            for key in keys {
+                if let webView = self.lookupWebViewHandlersByLookupSettings.removeValue(forKey: key)?.webViewHandler.webView {
+                    inMainThread {
+                        self.webViewProvider?.removeWebView(webView)
+                    }
+                }
+            }
+            self.remoteFileDownloader.stop()
+            let lookupData = self.lookupData
+            cleanupLookup(force: true, alreadyInQueue: true, update: Update(kind: .finishedAllLookups))
+            let storedItemResponses: [(ItemResponse, LibraryIdentifier)] = lookupData.compactMap {
+                switch $0.state {
+                case .translated(let translatedLookupData):
+                    return (translatedLookupData.response, translatedLookupData.libraryId)
+                    
+                default:
+                    return nil
+                }
+            }
+            for (response, libraryId) in storedItemResponses {
+                self.backgroundQueue.async { [weak self] in
+                    guard let self else { return }
+                    do {
+                        let request = MarkItemsAsTrashedDbRequest(keys: [response.key], libraryId: libraryId, trashed: true)
+                        try dbStorage.perform(request: request, on: backgroundQueue)
+                    } catch let error {
+                        DDLogError("IdentifierLookupController: can't trash item(s) - \(error)")
+                    }
+                }
+            }
+        }
+    }
+    
+    func getIdentifiersLookupCount(callback: @escaping (Int, Int, Int, [LookupData]) -> Void) {
+        accessQueue.async { [weak self] in
```

Comment:
Does this need to be asynchronous? If yes, this looks overly complicated, I'd just do
```
accessQueue.async { [weak self] in
    guard let self else { return }
    callback(self.lookupSavedCount, self.lookupFailedCount, self.lookupTotalCount, self.lookupData)
}
```

Ideally it would be synchronous and then you'd need all those variables.

Reply:
Check https://github.com/zotero/zotero-ios/pull/707/commits/7b9aa5de1ea91ae6227dbbbd0bd94261c2e0745d